File: Programming languages/parciales/parcial02/cajacambios/automatizador.cpp
#include "automatizador.h"
#include "cambios.h"
AutomatizadorCambios::AutomatizadorCambios(CajaDeCambios caja): caja(caja) {}

AutomatizadorCambios::~AutomatizadorCambios() {}
void neutralizar();
void reversar();
void subir();
void bajar();
void AutomatizadorCambios::neutralizar() {
  if (caja.obtenerCambio() == reversa) {
    bajar();
  } else {
    if (caja.obtenerCambio() != neutra) {
      caja.establecerEmbr();
      caja.ponerNeutra();
      caja.liberarEmb();
    }
  }
}

void AutomatizadorCambios::reversar() {
  if (caja.obtenerCambio() != reversa) {
    if (caja.obtenerCambio() != neutra) {
      neutralizar();
    }
    caja.establecerEmbr();
    caja.ponerReversa();
    caja.liberarEmb();
  }
}

void AutomatizadorCambios::subir() {
    if (caja.obtenerCambio() == neutra) {
      caja.establecerEmbr();
      caja.ponerPrimera();
      caja.liberarEmb();
    } else if (caja.obtenerCambio() == primera) {
      caja.establecerEmbr();
      caja.ponerSegunda();
      caja.liberarEmb();
    } else if (caja.obtenerCambio() == segunda) {
      caja.establecerEmbr();
      caja.ponerTercera();
      caja.liberarEmb();
	} else if (caja.obtenerCambio() == tercera) {
        caja.establecerEmbr();
        caja.ponerCuarta();
        caja.liberarEmb();
      } else if (caja.obtenerCambio() == cuarta) {
        caja.establecerEmbr();
        caja.ponerQuinta();
        caja.liberarEmb();
      }
    }

    void AutomatizadorCambios::bajar() {
        if (caja.obtenerCambio() == reversa) {
          caja.establecerEmbr();
          caja.ponerNeutra();
          caja.liberarEmb();
        } else if (caja.obtenerCambio() == primera) {
		  neutralizar();
        } else if (caja.obtenerCambio() == segunda) {
          caja.establecerEmbr();
          caja.ponerPrimera();
          caja.liberarEmb();
		}else if (caja.obtenerCambio() == tercera) {
            caja.establecerEmbr();
            caja.ponerSegunda();
            caja.liberarEmb();
          } else if (caja.obtenerCambio() == cuarta) {
            caja.establecerEmbr();
            caja.ponerTercera();
            caja.liberarEmb();
          } else if (caja.obtenerCambio() == quinta) {
            caja.establecerEmbr();
            caja.ponerCuarta();
            caja.liberarEmb();
          }
        }
	
         EstadoCambio AutomatizadorCambios::obtenerCambio() const {
          return caja.obtenerCambio();
        }

File: Programming languages/parciales/parcial03/cajacambios/automatizador.rb

class AutomatizadorCambios
	def initialize(caja)
		@caja = caja
	end
	def neutralizar
		if @caja.obtenerCambio == :reversa
			bajar
		elsif @caja.obtenerCambio != :neutra
			 @caja.establecerEmb
     		 @caja.ponerNeutra
   		     @caja.liberarEmb
		end
	end
	def reversar
		if @caja.obtenerCambio != :reversa 
    		if @caja.obtenerCambio != :neutra
     		 neutralizar
     		end
		     @caja.establecerEmb
   			 @caja.ponerReversa
   			 @caja.liberarEmb
   		end
  end

	def subir
		if @caja.obtenerCambio == :neutra
	    @caja.establecerEmb
        @caja.ponerPrimera
        @caja.liberarEmb
        elsif (@caja.obtenerCambio == :primera) 
        @caja.establecerEmb
        @caja.ponerSegunda
        @caja.liberarEmb
        elsif (@caja.obtenerCambio == :segunda)
        @caja.establecerEmb
        @caja.ponerTercera
        @caja.liberarEmb
	    elsif (@caja.obtenerCambio == :tercera) 
        @caja.establecerEmb
        @caja.ponerCuarta
        @caja.liberarEmb
        elsif (@caja.obtenerCambio == :cuarta) 
        @caja.establecerEmb
        @caja.ponerQuinta
        @caja.liberarEmb
      end
  end
  def bajar 
        if (@caja.obtenerCambio == :reversa) 
          @caja.establecerEmb
          @caja.ponerNeutra
          @caja.liberarEmb
        elsif (@caja.obtenerCambio == :primera) 
		  neutralizar
        elsif (@caja.obtenerCambio == :segunda) 
          @caja.establecerEmb
          @caja.ponerPrimera
          @caja.liberarEmb
		elsif (@caja.obtenerCambio == :tercera) 
            @caja.establecerEmb
            @caja.ponerSegunda
            @caja.liberarEmb
          elsif (@caja.obtenerCambio == :cuarta) 
            @caja.establecerEmb
            @caja.ponerTercera
            @caja.liberarEmb
          elsif (@caja.obtenerCambio == :quinta) 
            @caja.establecerEmb
            @caja.ponerCuarta
            @caja.liberarEmb
          end
   end
   def obtenerCambio
   	@caja.obtenerCambio
   end
end
